kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandDecryptionManager.java

```diff
@@ -0,0 +1,212 @@
+    }
+
+    private Set<E> extractEdeks(String topicName, int partition, MemoryRecords records) {
+        Set<E> edeks = new HashSet<>();
+        Serde<E> serde = kms.edekSerde();
+        for (Record kafkaRecord : records.records()) {
+            var decryptionVersion = decryptionVersion(topicName, partition, kafkaRecord);
+            if (decryptionVersion == EncryptionVersion.V1) {
+                ByteBuffer wrapper = kafkaRecord.value();
+                var edekLength = ByteUtils.readUnsignedVarint(wrapper);
+                ByteBuffer slice = wrapper.slice(wrapper.position(), edekLength);
+                var edek = serde.deserialize(slice);
+                edeks.add(edek);
+            }
+        }
+        return edeks;
+    }
+
+    @NonNull
+    private BatchAwareMemoryRecordsBuilder decrypt(String topicName,
+                                                   int partition,
+                                                   @NonNull MemoryRecords records,
+                                                   @NonNull BatchAwareMemoryRecordsBuilder builder,
+                                                   @NonNull Map<E, AesGcmEncryptor> encryptorMap,
+                                                   @NonNull List<Integer> batchRecordCounts) {
+        int i = 0;
+        for (MutableRecordBatch batch : records.batches()) {
+            Integer batchRecordCount = batchRecordCounts.get(i++);
+            if (batchRecordCount == 0 || batch.isControlBatch()) {
+                builder.writeBatch(batch);
+            }
+            else {
+                decryptBatch(topicName, partition, builder, encryptorMap, batch);
+            }
+        }
+        return builder;
+    }
+
+    private void decryptBatch(String topicName, int partition, @NonNull BatchAwareMemoryRecordsBuilder builder, @NonNull Map<E, AesGcmEncryptor> encryptorMap,
+                              MutableRecordBatch batch) {
+        builder.addBatchLike(batch);
+        for (Record kafkaRecord : batch) {
+            var decryptionVersion = decryptionVersion(topicName, partition, kafkaRecord);
+            if (decryptionVersion == null) {
+                builder.append(kafkaRecord);
+            }
+            else if (decryptionVersion == EncryptionVersion.V1) {
+                ByteBuffer wrapper = kafkaRecord.value();
+                var edekLength = ByteUtils.readUnsignedVarint(wrapper);
+                ByteBuffer slice = wrapper.slice(wrapper.position(), edekLength);
+                var edek = edekSerde.deserialize(slice);
+                wrapper.position(wrapper.position() + edekLength);
+                AesGcmEncryptor aesGcmEncryptor = encryptorMap.get(edek);
+                if (aesGcmEncryptor == null) {
+                    throw new EncryptionException("no encryptor loaded for edek, " + edek);
+                }
+                decryptRecord(EncryptionVersion.V1, aesGcmEncryptor, wrapper, kafkaRecord, builder);
+            }
```

Comment:
given we have an else-if tree that would fall through to skipping the record, should we throw here on an unexpected EncryptionVersion? Even better if we could have a parameterised test set up to fail as soon as V2 is added to EncryptionVersion.

Reply:
There are some changes in my PR for the decryption path that change how how the encryption versions stuff works. In order to avoid conflicts I'd prefer if we discuss and maybe change this at that point.